### ISSUE TITLE
package version

### ISSUE THREAD

**Issue by alllel** [mhiley/rust-netcdf#16]
The "crates.io" version of this package (0.1.0) is from 2015. And still the current version is 0.1.0

Maybe version in Cargo.toml should be increased? At least to 0.2.0?
There are new methods such as 'Variable::values_at', etc. compared to version of 2015.

**Comment by milesgranger** [comment]
Second this. :+1:

**Comment by evomassiny** [comment]
Me too :+1:

**Comment by milesgranger** [comment]
@mhiley What say you?

**Comment by mhiley** [comment]
Hi @milesgranger , I apologize for my lack of attention on your PR and these issues. Are you interested in becoming a collaborator on this repository? Currently myself and @evomassiny have push access, but I would be happy to add you. My interest in this project has waned, but I don't want that to stop you from pushing up new crate versions if you are actively developing.

**Comment by milesgranger** [comment]
Hello @mhiley , it (and life) happens, no worries.

I would be interested, but I'm concerned I'm a bit under-qualified. Discovered this while working on the PR, not sure I really understand the interfacing with C or even how the netcdf binding file(s) were created. Whilst I could spend time learning, I'm afraid I'd fall into the same trap, just not have time to properly dedicate to the repo. 

But with that said, if you won't completely abandon this, I could do my best to help out where I can, so long as you're available to ask the occasional question/clarification on work. :+1:

**Comment by mhiley** [comment]
Ok sounds good. I'll make an effort to re-immerse in this a bit and help keep you moving along. With regards to this specific issue - I see you bumped the package version in the PR that you closed due to lack of attention. Can you re-open that and I'll take a look, and we'll get that one merged?

**Comment by cbs228** [comment]
The version of this crate published on [crates.io](https://crates.io/crates/netcdf/0.1.0) has a heap corruption bug. This could cause memory corruption or undefined behavior. I would definitely recommend publishing an updated crate.

To answer the question that @mhiley raised in eff3d1330509a239fb126ed17ad2c9ba9b3f8714: the test was probably failing due to a use-after-free in `init_sub_groups()`. Please allow me explain this issue verbosely, as it is of interest to anyone interacting with C programs.

Before this commit, `groups.rs:243` read something like this:

```rust
let c_str: &ffi::CStr;
unsafe {
  // ...
  // name
  let mut buf_vec = vec![0i8; (namelen+1) as usize];
  let buf_ptr : *mut i8 = buf_vec.as_mut_ptr();
  // ...
  c_str = ffi::CStr::from_ptr(buf_ptr);
}
let str_buf: String = string_from_c_str(c_str); //<-- BUG!
```

`c_str` does not own the memory it references. Neither does `buf_ptr`. They both rely on a reference to `buf_vec`. At the end of the `unsafe` block, `buf_vec` passes out of scope, and its memory is released. This invalidates `buf_ptr` and `c_str`. Safe rust would enforce this lifetime, but unsafe rust does not.

`str_buf` is a `String` which owns its memory. Its memory is separate and safe... but it is constructed by copying a pointer that is no longer valid. This causes undefined behavior. Maybe `glibc` will notice and kill your program, or maybe you'll get invalid data. Or maybe—most unsettling—nothing will happen.

The fix was to assign to `str_buf` before `buf_vec` passes out of scope.

In general, it is probably a good idea to limit `unsafe` blocks to only code that is truly unsafe—i.e., won't build without it. It might be worth taking a look at some of the other `unsafe` blocks. Some of them could be made smaller or chopped up into several smaller blocks.

**Comment by mhiley** [comment]
closed by #21
